### PR TITLE
ISSUE-71 initial endpoint implementation for resource registration

### PR DESCRIPTION
**Pull request by sbryzak**
Still need to add some tests... this endpoint won't be fully functional until PAT support is implemented.

### CODE REVIEW

File: authorization/resource/resource_type.go [hunk 1]
```diff
@@ -77,6 +78,36 @@ func (m *GormResourceTypeRepository) Load(ctx context.Context, id uuid.UUID) (*R
...
+			ResourceTypeID: uuid.NewV4(),
+			Name:           name,
+		}
+
+		err := m.db.Create(native).Error
+
+		if err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"resource_type_id": native.ResourceTypeID,
+				"err":              err,
```
Comment (alexeykazakov) [inline authorization/resource/resource_type.go]: let's add "name": name too

File: authorization/resource/resource_type.go [hunk 2]
```diff
@@ -77,6 +78,36 @@ func (m *GormResourceTypeRepository) Load(ctx context.Context, id uuid.UUID) (*R
...
+			}, "unable to create the resource type")
+			return nil, errs.WithStack(err)
+		}
+		log.Debug(ctx, map[string]interface{}{
+			"resource_type_id": native.ResourceTypeID,
```
Comment (alexeykazakov) [inline authorization/resource/resource_type.go]: let's add "name": name too

File: controller/resource.go [hunk 3]
```diff
@@ -37,12 +43,57 @@ func (c *ResourceController) Read(ctx *app.ReadResourceContext) error {
...
+	// TODO validate the PAT here
 
-	// ResourceController_Register: end_implement
-	return nil
+	var res *resource.Resource
+
+	err := application.Transactional(c.db, func(appl application.Application) error {
+
+		// Lookup or create the resource type
+		resourceType, err := appl.ResourceTypeRepository().LookupOrCreate(ctx, ctx.Payload.Name)
```
Comment (alexeykazakov) [inline controller/resource.go]: Will we allow to create resource type by resource servers? Or should we pre-define them?

File: controller/resource.go [hunk 4]
```diff
@@ -37,12 +43,57 @@ func (c *ResourceController) Read(ctx *app.ReadResourceContext) error {
...
+			return jsonapi.JSONErrorResponse(ctx, err)
+		}
+
+		// Lookup the parent resource if one has been specified
+		var parentResource *resource.Resource
+		if (ctx.Payload.ParentResourceID != nil) {
+			parentResource, err = appl.ResourceRepository().Load(ctx, *ctx.Payload.ParentResourceID)
+
+			if (err != nil) {
+				return jsonapi.JSONErrorResponse(ctx, err)
```
Comment (alexeykazakov) [inline controller/resource.go]: Just return err here. You will wrap it as JSONErrorResponse outside of the transaction (see line 93)

File: controller/resource.go (hunk 3 above)
Comment (sbryzak) [reply controller/resource.go]: I think it's easiest to just allow the resource servers to control this, no need for us to pre-define the resource types.  Also this makes it compliant with the UMA spec.

File: authorization/resource/resource_type.go (hunk 2 above)
Comment (sbryzak) [reply authorization/resource/resource_type.go]: I've added the "name" fields to the logging.

File: controller/resource.go [hunk 5]
```diff
@@ -37,12 +44,67 @@ func (c *ResourceController) Read(ctx *app.ReadResourceContext) error {
...
-	// ResourceController_Register: end_implement
-	return nil
+	var res *resource.Resource
+
+	err := application.Transactional(c.db, func(appl application.Application) error {
+
+		// Lookup or create the resource type
+		resourceType, err := appl.ResourceTypeRepository().LookupOrCreate(ctx, ctx.Payload.Type)
+		if (err != nil) {
+			return jsonapi.JSONErrorResponse(ctx, err)
```
Comment (alexeykazakov) [inline controller/resource.go]: return err

File: controller/resource.go [hunk 6]
```diff
@@ -37,12 +44,67 @@ func (c *ResourceController) Read(ctx *app.ReadResourceContext) error {
...
+		}
+
+		// TODO extract the owner identity from the PAT
+		var identityID uuid.UUID
+
+		identity, err := appl.Identities().Load(ctx, identityID)
+		if (err != nil) {
+			// TODO raise an error if the identity could not be determined
+
+			return jsonapi.JSONErrorResponse(ctx, err)
```
Comment (alexeykazakov) [inline controller/resource.go]: return err

File: controller/resource.go [hunk 7]
```diff
@@ -37,12 +44,67 @@ func (c *ResourceController) Read(ctx *app.ReadResourceContext) error {
...
+	})
+
+
+	if (err != nil) {
+	  return jsonapi.JSONErrorResponse(ctx, err)
+	}
+
+	log.Debug(ctx, map[string]interface{}{
+		"resource_id":        res.ResourceID,
+    "parent_resource_id": res.ParentResource.ResourceID,
```
Comment (alexeykazakov) [inline controller/resource.go]: `make format-go-code` ;)
Btw you can configure Gogland to do `go fmt` when saving the file

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-94-1`

**Comment by sbryzak** [comment]
I've added the start of the endpoint tests for resource registration, however I'm getting the following weird error which I can't seem to fix:

=== RUN   TestFailRegisterResourceUnauthorized
time="2017-09-14T17:19:47+10:00" level=error msg="Resource owner ID is not valid" file="/home/shane/go/src/github.com/fabric8-services/fabric8-auth/controller/resource.go" func="github.com/fabric8-services/fabric8-auth/controller.(*ResourceController).Register.func1" line=73 pid=22310 pkg="controller.(*ResourceController).Register" req_id= resource_owner_id=xxx1234 
time="2017-09-14T17:19:47+10:00" level=info msg="an error occurred in our api" err="[97cZ3Jss] 401 unauthorized: Invalid resource_owner_id" error_message="[97cZ3Jss] 401 unauthorized: Invalid resource_owner_id" pkg=jsonapi req_id= 

[35m(unsupported destination, should be slice or struct)[0m 
[33m[2017-09-14 17:19:47][0m [31;1m [0m
[info] removing callback `mighti:record` from /home/shane/go/src/github.com/fabric8-services/fabric8-auth/gormsupport/cleaner/db_clean.go:78
--- FAIL: TestFailRegisterResourceUnauthorized (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x98fd47]

It's probably something I've done wrong somewhere, could one of you guys please take a look at my test and see if anything obvious stands out?  Thanks!

**Comment by alexeykazakov** [comment]
https://github.com/sbryzak/fabric8-auth/pull/1

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-94-2`

File: controller/resource.go [hunk 8]
```diff
@@ -42,11 +44,16 @@ func (c *ResourceController) Read(ctx *app.ReadResourceContext) error {
 // Register runs the register action.
 func (c *ResourceController) Register(ctx *app.RegisterResourceContext) error {
 
-	// TODO validate the PAT here
+	// Validate the PAT
+	_, err := c.TokenManager.Locate(ctx)
+	if err != nil {
+		jerrors, _ := jsonapi.ErrorToJSONAPIErrors(ctx, goa.ErrBadRequest(err.Error()))
+		return ctx.BadRequest(jerrors)
```
Comment (sbose78) [inline controller/resource.go]: We have started using the following format everywhere
`jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError(...pass the err object))`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-94-10`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-94-11`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-94-12`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-94-13`

File: controller/resource_blackbox_test.go [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+		ResourceOwnerID:  resourceOwnerID.String(),
+		Type:             "Area",
+	}
+
+	test.RegisterResourceBadRequest(rest.T(), service.Context, service, controller, payload)
+}
+
+func (rest *TestResourceREST) TestRegisterResourceCreated() {
+
+	testIdentity, err := testsupport.CreateTestIdentity(rest.DB, "TestRegisterResourceCreated-"+uuid.NewV4().String(), "TestRegisterResourceCreated")
```
Comment (alexeykazakov) [inline controller/resource_blackbox_test.go]: You don't need to create an identity in the DB to use it as a service account, do you?
And it's better to create one regular identity and one service account in SetupSuite() method so they are available for all tests and you can reuse them. You could avoid copy-paste the same code in every test method then.

File: controller/resource_blackbox_test.go [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+		Username: "fabric8-wit",
+	}
+	service, controller := rest.SecuredControllerWithServiceAccount(sa)
+
+	resourceDescription := "Resource description"
+	resourceID := ""
+	resourceScopes := []string{}
+	resourceOwnerID := testIdentity.ID
+
+	fmt.Printf("!!!!!! ID: %v\n", resourceOwnerID)
```
Comment (alexeykazakov) [inline controller/resource_blackbox_test.go]: Remove printf.

File: controller/resource_blackbox_test.go [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+		Description:      &resourceDescription,
+		Name:             "My new resource",
+		ParentResourceID: nil,
+		ResourceScopes:   resourceScopes,
+		ResourceID:       &resourceID,
+		ResourceOwnerID:  resourceOwnerID.String(),
+		Type:             "Area",
+	}
+
+	fmt.Println("Creating...")
```
Comment (alexeykazakov) [inline controller/resource_blackbox_test.go]: Remove println.

File: controller/resource_blackbox_test.go [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+	_, created := test.RegisterResourceCreated(rest.T(), service.Context, service, controller, payload)
+	// then
+	fmt.Println("...Created")
```
Comment (alexeykazakov) [inline controller/resource_blackbox_test.go]: Remove println.

File: controller/resource_blackbox_test.go [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+		Username: "fabric8-wit",
+	}
+	service, controller := rest.SecuredControllerWithServiceAccount(sa)
+
+	resourceDescription := "Resource description"
+	resourceID := uuid.NewV4().String()
+	resourceScopes := []string{}
+	resourceOwnerID := testIdentity.ID
+
+	fmt.Printf("!!!!!! ID: %v\n", resourceOwnerID)
```
Comment (alexeykazakov) [inline controller/resource_blackbox_test.go]: Remove println.

File: controller/resource_blackbox_test.go [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+		Description:      &resourceDescription,
+		Name:             "My new resource",
+		ParentResourceID: nil,
+		ResourceScopes:   resourceScopes,
+		ResourceID:       &resourceID,
+		ResourceOwnerID:  resourceOwnerID.String(),
+		Type:             "Area",
+	}
+
+	fmt.Println("Creating...")
```
Comment (alexeykazakov) [inline controller/resource_blackbox_test.go]: Remove println.

File: token/token.go [hunk 15]
```diff
@@ -454,6 +454,13 @@ func IsSpecificServiceAccount(ctx context.Context, names []string) bool {
 	return false
 }
 
+// IsServiceAccount checks if the request is done by a
+// Service account based on the JWT Token provided in context
+func IsServiceAccount(ctx context.Context) bool {
```
Comment (alexeykazakov) [inline token/token.go]: Add a test for this function.

Comment (sbryzak) [reply token/token.go]: I've added a couple of tests for this method to the token test suite.

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-94-14`

**Comment by fabric8cd** [comment]
@sbryzak snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-94-15`

**Pull request merged**